p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,790 @@
+	// client round tripper in a thread-safe way.
+	createDefaultClientRoundTripper sync.Once
+	httpTransport                   *httpTransport
+}
+
+type httpTransport struct {
+	listenAddrs         []ma.Multiaddr
+	listeners           []net.Listener
+	closeListeners      chan struct{}
+	waitingForListeners chan struct{}
+}
+
+func newPeerMetadataCache() *lru.Cache[peer.ID, PeerMeta] {
+	peerMetadata, err := lru.New[peer.ID, PeerMeta](peerMetadataLRUSize)
+	if err != nil {
+		// Only happens if size is < 1. We make sure to not do that, so this should never happen.
+		panic(err)
+	}
+	return peerMetadata
+}
+
+func (h *Host) httpTransportInit() {
+	h.createHTTPTransport.Do(func() {
+		h.httpTransport = &httpTransport{
+			closeListeners:      make(chan struct{}),
+			waitingForListeners: make(chan struct{}),
+		}
+	})
+}
+
+func (h *Host) serveMuxInit() {
+	h.initializeServeMux.Do(func() {
+		h.ServeMux = http.NewServeMux()
+	})
+}
+
+func (h *Host) Addrs() []ma.Multiaddr {
+	h.httpTransportInit()
+	<-h.httpTransport.waitingForListeners
+	return h.httpTransport.listenAddrs
+}
+
+// ID returns the peer ID of the underlying stream host, or the zero value if there is no stream host.
+func (h *Host) PeerID() peer.ID {
+	if h.StreamHost != nil {
+		return h.StreamHost.ID()
+	}
+	return ""
+}
+
+var ErrNoListeners = errors.New("nothing to listen on")
+
+// Serve starts the HTTP transport listeners. Always returns a non-nil error.
+// If there are no listeners, returns ErrNoListeners.
+func (h *Host) Serve() error {
+	// assert that each addr contains a /http component
+	for _, addr := range h.ListenAddrs {
+		_, isHTTP := normalizeHTTPMultiaddr(addr)
+		if !isHTTP {
+			return fmt.Errorf("address %s does not contain a /http or /https component", addr)
+		}
+	}
+
+	h.serveMuxInit()
+	h.ServeMux.Handle("/.well-known/libp2p", &h.WellKnownHandler)
+
+	h.httpTransportInit()
+
+	closedWaitingForListeners := false
+	defer func() {
+		if !closedWaitingForListeners {
+			close(h.httpTransport.waitingForListeners)
+		}
+	}()
+
+	if len(h.ListenAddrs) == 0 && h.StreamHost == nil {
+		return ErrNoListeners
+	}
+
+	h.httpTransport.listeners = make([]net.Listener, 0, len(h.ListenAddrs)+1) // +1 for stream host
+
+	streamHostAddrsCount := 0
+	if h.StreamHost != nil {
+		streamHostAddrsCount = len(h.StreamHost.Addrs())
+	}
+	h.httpTransport.listenAddrs = make([]ma.Multiaddr, 0, len(h.ListenAddrs)+streamHostAddrsCount)
+
+	errCh := make(chan error)
+
+	if h.StreamHost != nil {
+		listener, err := streamHostListen(h.StreamHost)
+		if err != nil {
+			return err
+		}
+		h.httpTransport.listeners = append(h.httpTransport.listeners, listener)
+		h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, h.StreamHost.Addrs()...)
+
+		go func() {
+			errCh <- http.Serve(listener, h.ServeMux)
+		}()
+	}
+
+	closeAllListeners := func() {
+		for _, l := range h.httpTransport.listeners {
+			l.Close()
+		}
+	}
+
+	err := func() error {
```

Comment:
Maybe cleaner: Make this a separate method that returns the listen addrs and an error.

Reply:
This also adds to the listeners. Maybe I'll have this accept a pointer of http transport so it's clear what is getting mutated.

Reply:
Actually moved this out into a separate method so it was clear what is in scope vs not.